xcpmd/src/battery.c

```diff
@@ -0,0 +1,803 @@
+    }
+    else if (!strcmp(attrib_name, "energy_full_design")) {
+        info->energy_full_design = strtoull(attrib_value, NULL, 10) / 1000;
+    }
+    else if (!strcmp(attrib_name, "energy_full")) {
+        info->energy_full = strtoull(attrib_value, NULL, 10) / 1000;
+    }
+    else if (!strcmp(attrib_name, "voltage_min_design")) {
+        info->design_voltage = strtoull(attrib_value, NULL, 10) / 1000;
+    }
+    else if (!strcmp(attrib_name, "model_name")) {
+        strncpy(info->model_number, attrib_value, 32);
+    }
+    else if (!strcmp(attrib_name, "serial_number")) {
+        strncpy(info->serial_number, attrib_value, 32);
+    }
+    else if (!strcmp(attrib_name, "technology")) {
+        if (strstr(attrib_value, "Li-ion"))
+	        strncpy(info->battery_type, "LION\n\0", 6);
+	    else if (strstr(attrib_value, "Li-poly"))
+	        strncpy(info->battery_type, "LiP\n\0", 6);
+	    else
+	        strncpy(info->battery_type, attrib_value, 32);
+        info->battery_technology = RECHARGEABLE;
+    }
+    else if (!strcmp(attrib_name, "manufacturer")) {
+      strncpy(info->oem_info, attrib_value, 32);
+    }
+}
+
+
+//Gets a battery's status from the sysfs and stores it in last_status.
+int update_battery_status(unsigned int battery_index) {
+
+    int i, rc;
+    DIR *battery_dir;
+    struct dirent * dp;
+    FILE *file;
+    char filename[256];
+    char data[128];
+    char *ptr;
+
+    struct battery_status status;
+    memset(&status, 0, sizeof(struct battery_status));
+
+    if (battery_index >= num_battery_structs_allocd)
+        return -1;
+
+    if (battery_slot_exists(battery_index) == NO) {
+        status.present = NO;
+        memcpy(&last_status[battery_index], &status, sizeof(struct battery_status));
+        return 1;
+    }
+
+    battery_dir = get_battery_dir(battery_index);
+    if (!battery_dir) {
+        //Battery directory does not exist--this normally occurs when a battery slot is removed
+        if (errno == ENOENT) {
+            status.present = NO;
+            //memcpy(&last_status[battery_index], &status, sizeof(struct battery_status));
+            return 1;
+        }
+        else {
+            xcpmd_log(LOG_ERR, "opendir in update_battery_status for directory %s/BAT%d failed with error %d\n", BATTERY_DIR_PATH, battery_index, errno);
+            return 0;
+        }
+    }
+
+    //Loop over the files in the directory.
+    while ((dp = readdir(battery_dir)) != NULL) {
+
+        //Convert from dirent to file and read out the data.
+        if (dp->d_type == DT_REG) {
+
+            memset(filename, 0, sizeof(filename));
+            sprintf(filename, "%s/BAT%i/%s", BATTERY_DIR_PATH, battery_index, dp->d_name);
+
+            file = fopen(filename, "r");
+            if (file == NULL)
+                continue;
+
+            memset(data, 0, sizeof(data));
+            fgets(data, sizeof(data), file);
+
+            fclose(file);
+
+            //Trim off leading spaces.
```

Comment:
Oh, that must come from the time we were using /proc!
I guess it doesn't hurt to keep doing it.